### DISCUSSION TITLE
Build failing when middleware is present

### DISCUSSION THREAD

**Discussion by jakubschenk** [vercel/community#1321] [Help] [answered]
### Summary

Hi,
I have a project with Next@12 and Storybook installed. I have wanted to migrate to Next@13, but this breaks my Storybook deployment from the same repository. The issue seems to be with the "Other" config for the Storybook deployment, which crashes on `Cannot import module 'crypto'`, because build api tries to pick up `middleware.ts` from the root of the project, which happened after moving the file from `pages/_middleware.ts` to `middleware.ts` after [Migration to Next13](https://nextjs.org/docs/messages/middleware-upgrade-guide).

Build command: npm run build-storybook
Build directory: storybook-static
Install command: npm ci

Is there a way to make Vercel Deployment ignore the middleware or any other way? I have tried just rimraf-ing the file before building bu that fails.

Thanks for any information.

### Additional information

The project deployments for the Next app itself deploy successfully.
`Error: The Edge Function "middleware" is referencing unsupported modules: crypto, crypto, crypto`

### Example

https://github.com/jakubschenk/storybook-deploy-example-fail

**Comment by Zebeqo** [comment]
This happened to me too. A next.13 appDir project from scratch.

**Reply by korsvanloon** [reply]
I had the same, then removed the crypto dependency form my middleware and it still breaks:

```
ReferenceError: __dirname is not defined
    at (node_modules/.pnpm/next@13.1.6_m5sxuueb27gk6ddc5gums6vtgq/node_modules/next/dist/compiled/ua-parser-js/ua-parser.js:10:16756)
    at (node_modules/.pnpm/next@13.1.6_m5sxuueb27gk6ddc5gums6vtgq/node_modules/next/dist/compiled/ua-parser-js/ua-parser.js:10:16885)
    at (node_modules/.pnpm/next@13.1.6_m5sxuueb27gk6ddc5gums6vtgq/node_modules/next/dist/server/web/spec-extension/user-agent.js:8:41)
    at (node_modules/.pnpm/next@13.1.6_m5sxuueb27gk6ddc5gums6vtgq/node_modules/next/server.js:6:0)
    at (middleware.js:38:29)
    at (middleware:middleware.js:1:17)
```

Apparently, having any middleware at all breaks (that is not used by storybook of course) breaks the build.

**Comment by ewgenius** [comment]
Same problem on next@12.3.4 - I have `middleware.ts` at the root of the project and Vercel detects it during storybook build and fails because of that. I guess that it doesn't depend on next version.

**Reply by jakubschenk** [reply]
@ewgenius the version depends since the middleware location has changed in Next 12.2, which is the reason of the issue. On versions before that with the middleware in `pages/_middleware.{ts,js}` everything was fine.

**Reply by ewgenius** [reply]
got it, thanks for details

**Comment by amyegan** [comment]
The engineering team is exploring potential solutions. Thank you all for reaching out and sharing details about this issue. It has been very helpful for debugging 🙇

**Reply by sumanta-pramanick** [reply]
Hi @amyegan any update on this. I am also facing same issue. I can also share code snippet as well if required.

**Comment by andyWynyard** [comment]
Yep, we have had this issue in our project since adding middleware a few months ago. I just came back to this to fix it, but it seems it's a known issue. Great that you guys are on it.

**Comment by smaeda-ks** [comment, marked as answer]
There's an open PR to address this problem:
https://github.com/vercel/vercel/pull/9587

Once it's released, you will be able to mitigate this by dynamically updating the `.vercelignore` file at build time, as proposed in the PR description above.

**Reply by jmduke** [reply]
It looks like this PR has been closed; what is the right recourse here?

**Reply by ovflowd** [reply]
Same here, also curious about the state of this, @smaeda-ks

**Reply by smaeda-ks** [reply]
We've been working on it: https://github.com/vercel/vercel/pull/9839

**Reply by smaeda-ks** [reply]
Vercel CLI 29.1.1 has been released with support for Storybook projects. The Storybook framework can now be selected under the project settings.
